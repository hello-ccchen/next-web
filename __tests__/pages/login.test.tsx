/**
 * @jest-environment jsdom
 */
import { fireEvent, render } from "@testing-library/react";
import Login from "@/app/login/page";
import { useSearchParams } from "next/navigation";
import { signIn } from "next-auth/react";

jest.mock("next/navigation", () => ({
  useSearchParams: jest.fn(),
}));

jest.mock("next-auth/react", () => ({
  signIn: jest.fn(),
}));

describe("Login Page", () => {
  it("renders no access rigth when detected callbackurl", () => {
    // Arrange
    (useSearchParams as jest.Mock).mockReturnValue({
      get: () => "abcde",
    });

    // Act
    const { container, getByText } = render(<Login />);

    // Assert
    expect(container).toMatchSnapshot();
    expect(getByText("401: Uh-oh, you do not have access. 😟")).toBeTruthy();
  });

  it("renders `You're almost there!` when detected no callbackurl", () => {
    // Arrange
    (useSearchParams as jest.Mock).mockReturnValue({
      get: () => null,
    });

    // Act
    const { container, getByText } = render(<Login />);

    // Assert
    expect(container).toMatchSnapshot();
    expect(getByText(`You're almost there! 😁`)).toBeTruthy();
  });

  it("should sign in when user click the sign in button", () => {
    // Arrange
    (useSearchParams as jest.Mock).mockReturnValue({
      get: () => null,
    });

    // Act
    const { container } = render(<Login />);
    const loginButton = container.querySelector(".btn");
    expect(loginButton).toBeTruthy();
    fireEvent.click(loginButton as Element);

    // Assert
    expect(signIn).toHaveBeenCalledWith("google", { callbackUrl: "/" });
  });
});
