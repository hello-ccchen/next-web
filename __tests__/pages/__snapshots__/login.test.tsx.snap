// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page renders \`You're almost there!\` when detected no callbackurl 1`] = `
<div>
  <div
    class="container shadow-sm container"
  >
    <div
      class="row content"
    >
      <div
        class="bggradient bg-primary 
          col-sm-6 border d-flex justify-content-center align-items-center"
      >
        <h1
          class="text-white fw-bolder"
        >
          You're almost there! 😁
        </h1>
      </div>
      <div
        class="col-sm-6 border d-flex justify-content-center align-items-center"
      >
        <div
          class="d-flex flex-nowrap justify-content-center w-75 border shadow py-3 btn"
        >
          <img
            alt="google"
            class="me-2"
            data-nimg="1"
            decoding="async"
            height="24"
            loading="lazy"
            src="/google.svg"
            style="color: transparent;"
            width="24"
          />
          Continue with Google
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Login Page renders no access rigth when detected callbackurl 1`] = `
<div>
  <div
    class="container shadow-sm container"
  >
    <div
      class="row content"
    >
      <div
        class="bggradient bg-info 
          col-sm-6 border d-flex justify-content-center align-items-center"
      >
        <h1
          class="text-white fw-bolder"
        >
          401: Uh-oh, you do not have access. 😟
        </h1>
      </div>
      <div
        class="col-sm-6 border d-flex justify-content-center align-items-center"
      >
        <div
          class="d-flex flex-nowrap justify-content-center w-75 border shadow py-3 btn"
        >
          <img
            alt="google"
            class="me-2"
            data-nimg="1"
            decoding="async"
            height="24"
            loading="lazy"
            src="/google.svg"
            style="color: transparent;"
            width="24"
          />
          Continue with Google
        </div>
      </div>
    </div>
  </div>
</div>
`;
